### PR TITLE
Fixing race issues in log rotator

### LINKED ISSUE #1042: raw_exec driver is failing with: Wait returned exit code 141, signal 13, and error <nil>
### Nomad version
Nomad v0.3.2-dev ('a1996403081101ab952de12b7894a0f2b1716b8f')
Nomad v0.3.1

### Operating system and Environment details
Linux

### Issue
After app is deployed and some traffic goes through it, it is terminated and nomad is restarting it after few minutes. If app is running normally everything is working like a charm. 

```
2016/04/06 14:10:22 [INFO] client: task "tcpproxy-test" for alloc "ff8c8915-08e6-d981-d4d1-c949110e03ed" failed: Wait returned exit code 141, signal 13, and error <nil>
2016/04/06 14:10:22 [INFO] client: Restarting task "tcpproxy-test" for alloc "ff8c8915-08e6-d981-d4d1-c949110e03ed" in 16.672584629s
2016/04/06 14:10:22 [DEBUG] client: updated allocations at index 12378 (pulled 0) (filtered 2)
2016/04/06 14:10:22 [DEBUG] client: allocs: (added 0) (removed 0) (updated 0) (ignore 2)
```

Note that logs stderr and stdout are not containing any crash information. 

### Job
```
job "tcpproxy-test" {
    region = "xxx"
    datacenters = ["yyy"]
    type = "service"
    priority = 50

    update {
        # Stagger updates every 10 seconds
        stagger = "10s"

        # Update a single task at a time
        max_parallel = 1
    }

    group "tcpproxy-test" {
        count = 1
        # Define a task to run
        task "tcpproxy-test" {
            driver = "raw_exec"
            
            config {
                command = "tcpproxy"
                args = ["-h", ":8885", "-l",":2090", "-f", "192.168.0.4:8887"]
            }

            artifact {
                source = "https://someurltotcpproxy"
            }
    
            resources {
                cpu = 500 # 500 Mhz
                memory = 128
                disk = 100
            }
            
            logs {
                max_files = 5
                max_file_size = 10
            }
        }
    }
}
```
Related to #955

### PR DESCRIPTION
**Pull request by diptanu**
Fixed #1042

### CODE REVIEW

File: client/driver/logging/rotator.go [hunk 1]
```diff
@@ -88,12 +91,10 @@ func (f *FileRotator) Write(p []byte) (n int, err error) {
 		if remainingSize < int64(len(p[n:])) {
```
Comment (sean-) [inline client/driver/logging/rotator.go]: This should scan backwards through the buffer for a `\n` and write/flush everything before that.  Whatever comes after that last `\n` should be moved to the beginning of the buffer and written out on the next write (as long as the line length is less than the buffer length, in which case you need to truncate).  This only needs to happen on file rotation, not on write.  I'll open an issue re: this in a sec.

**Comment by sean-** [comment]
Opening an issue for the remaining item, but LGTM otherwise.

**Comment by a86c6f7964** [comment]
Just curious. Was the race in this just between the `flushPeriodically` and the `Write`? Y'all are assuming that `Write` is only called in sequence because it is reading the channel from syslog correct?

**Comment by dadgar** [comment]
@a86c6f7964 correct!

**Pull request merged**